### PR TITLE
fix(system proxy): app needs restart to fetch system proxy when set from terminal

### LINKED ISSUE #7559: System proxy changes require app restart to take effect
### I have checked the following:

- [x] I have searched existing issues and found nothing related to my issue.

### This bug is:

- [ ] making Bruno unusable for me
- [x] slowing me down but I'm able to continue working
- [x] annoying
- [ ] this feature was working in a previous version but is broken in the current release.

### Bruno version

3.2.0

### Operating System

Mac OS

### Describe the bug

When using system proxy variables, Bruno does not pick up changes dynamically. Every time proxy variables are set or modified, the application needs to be restarted for the changes to take effect.

This behavior is inconvenient, especially for users who frequently switch or update proxy configurations during development.

### Steps to reproduce

1. Set system proxy variables 

```bash
export HTTP_PROXY="http://127.0.0.1:8080"
export HTTPS_PROXY="http://127.0.0.1:8080"
```
2. Open Bruno and configure system proxy. 

3.  It won't update the values automatically

```bash
export HTTP_PROXY="http://127.0.0.1:9090"
export HTTPS_PROXY="http://127.0.0.1:9090"
```

4. Click the refresh button inside Bruno and it won't work.

5. Restart the app, and Bruno will update the values.

### Expected Behavior

- Bruno should automatically detect and apply updated system proxy variables

- The refresh action should reload and reflect the latest proxy configuration

### .bru file to reproduce the bug

_No response_

### Screenshots/Live demo link

Image

### PR DESCRIPTION
**Pull request by abhishekp-bruno**
JIRA Link - https://usebruno.atlassian.net/browse/BRU-3202

Fixes: https://github.com/usebruno/bruno/issues/7559

### Description

When users configured proxy settings via shell environment variables (e.g. `http_proxy`, `https_proxy` in `.zshrc`/`.bashrc`), Bruno would only pick them up at startup. Changing or removing those exports in the terminal had no effect until the app was fully restarted — the cached `process.env` proxy values went
stale, and the "refresh" path never re-read the shell.

### Fix

Added `refreshShellEnvProxyVars()` in `bruno-requests`, which re-syncs proxy-related variables from the user's login shell on demand:

- **Clears stale proxy vars first** (`delete process.env[key]`) before spawning the login shell, so a removed export in `.zshrc` doesn't persie subprocess.
- Re-fetches the shell environment and re-applies any proxy values that are still set.
- Covers the full set of proxy keys: `http_proxy`/`HTTP_PROXY`, `https_proxy`/`HTTPS_PROXY`, `no_proxy`/`NO_PROXY`, `all_proxy`/`ALL_PROXY`.

Wired this into the Electron system-proxy store: `fetchSystemProxy({ refresh: true })` now passes `refreshShellEnv: true` down to `loadSystemP re-reads the terminal environment rather than serving cached values.

### Changes

- `packages/bruno-requests/src/utils/shell-env.ts` — new `refreshShellEnvProxyVars()` + exported `PROXY_ENV_KEYS`.
- `packages/bruno-requests/src/index.ts` — export the new function.
- `packages/bruno-electron/src/store/system-proxy.js` — refresh shell env on proxy refresh.

### Testing

Added unit tests covering the refresh behavior (re-syncing updated vars and clearing removed vars).

#### Contribution Checklist:

- [x] **I've used AI significantly to create this pull request**
- [ ] **The pull request only addresses one issue or adds one feature.**
- [x] **The pull request does not introduce any breaking changes**
- [ ] **I have added screenshots or gifs to help explain the change if applicable.**
- [x] **I have read the [contribution guidelines](https://github.com/usebruno/bruno/blob/main/contributing.md).**
- [ ] **Create an issue and link to the pull request.**

Note: Keeping the PR small and focused helps make it easier to review and merge. If you have multiple changes you want to make, please consider submitting them as separate pull requests.

#### Publishing to New Package Managers

Please see [here](../publishing.md) for more information.

## Summary by CodeRabbit

* **Improvements**
  * System proxy reloads can now refresh shell environment proxy variables first, ensuring proxy settings stay in sync across the app.
  * Reload behavior now aligns with the existing “refresh” toggle, updating cached proxy values consistently.

* **Public API**
  * Exposed a capability to refresh shell-environment proxy variables programmatically.
  * Added a shared list of supported proxy environment variable keys.

* **Tests**
  * Added Jest coverage to verify proxy env var updates, removal when absent, and cleanup of stale lowercase/uppercase proxy variables.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -4949,7 +4949,7 @@
       "version": "7.26.3",
       "resolved": "https://registry.npmjs.org/@babel/helper-create-regexp-features-plugin/-/helper-create-regexp-features-plugin-7.26.3.tgz",
       "integrity": "sha512-G7ZRb40uUgdKOQqPLjfD12ZmGA54PzqDFUv2BKImnC9QIfGhIHKvVML0oN8IUiDq4iRqpq74ABpvOaerfWdong==",
-      "devOptional": true,
```
Comment (sanish-bruno) [inline package-lock.json]: rm these changes

File: packages/bruno-requests/src/utils/shell-env.ts [hunk 2]
```diff
@@ -35,3 +45,30 @@ export const initializeShellEnv = async (): Promise<Record<string, string>> => {
...
+  }
+
+  const shellEnvVars = await fetchShellEnv();
+
+  for (const key of PROXY_ENV_KEYS) {
+    const value = shellEnvVars[key];
+    if (value) {
+      process.env[key] = value;
+    } else {
+      delete process.env[key];
```
Comment (sanish-bruno) [inline packages/bruno-requests/src/utils/shell-env.ts]: the delete seems to be redundant, we are already deleting them above,if block will be enough!

File: packages/bruno-requests/src/utils/shell-env.ts [hunk 3]
```diff
@@ -35,3 +45,30 @@ export const initializeShellEnv = async (): Promise<Record<string, string>> => {
...
   return shellEnvVars;
 };
+
+/**
+ * Re-syncs proxy-related process.env values from the user's shell configuration.
+ * Used when refreshing system proxy settings without restarting the app.
+ *
+ * @returns The fetched shell environment variables
+ */
+export const refreshShellEnvProxyVars = async (): Promise<Record<string, string>> => {
```
Comment (sanish-bruno) [inline packages/bruno-requests/src/utils/shell-env.ts:86]: should we have early return for windows platforms

```js
if (process.platform === 'win32') {
    return {};
  }
```

**Review by sanish-bruno — CHANGES_REQUESTED** [review]

File: packages/bruno-requests/src/utils/shell-env.ts (hunk 2 above)
Comment (abhishekp-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts]: REMOVED

**Review by sanish-bruno — APPROVED** [review]

File: packages/bruno-electron/src/store/system-proxy.js [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
+const { getSystemProxy, refreshShellEnvProxyVars } = require('@usebruno/requests');
 
 let cachedSystemProxy;
 let systemProxyPromise;
 
-const loadSystemProxy = async () => {
+const loadSystemProxy = async ({ refreshShellEnv = false } = {}) => {
   try {
+    if (refreshShellEnv) {
+      await refreshShellEnvProxyVars();
```
Comment (bijin-bruno) [inline packages/bruno-electron/src/store/system-proxy.js:9]: Have we made sure this is not blocking for the systems that doesn't have access to shellEnv? 
Ref: 
1. code -> `packages/bruno-electron/src/store/shell-env-state.js`
2. prev issue -> https://github.com/usebruno/bruno/issues/7568

**Review by bijin-bruno — CHANGES_REQUESTED** [review]

File: packages/bruno-electron/src/store/system-proxy.js (hunk 4 above)
Comment (abhishekp-bruno) [reply packages/bruno-electron/src/store/system-proxy.js:9]: Have added a timeout to refresh to handle this issue

**Review by lohit-bruno — DISMISSED** [review]

File: packages/bruno-requests/src/utils/shell-env.spec.ts [hunk 5]
```diff
@@ -92,3 +92,54 @@ describe('initializeShellEnv', () => {
...
+      http_proxy: 'http://new-proxy:8080',
+      https_proxy: 'http://new-proxy:8443'
+    };
+
+    await refreshShellEnvProxyVars();
+
+    expect(process.env.http_proxy).toBe('http://new-proxy:8080');
+    expect(process.env.https_proxy).toBe('http://new-proxy:8443');
+  });
+
```
Comment (helloanoop) [inline packages/bruno-requests/src/utils/shell-env.spec.ts]: The tests appear to be achieveing the same thing
* `should remove proxy env vars missing from shell config`
* `should not keep stale process.env proxy vars after shell config is cleared`

both tests set `mockShellEnvResult = {};` and the check the process env vars to be undefined, seems to be a bit redundant

File: packages/bruno-requests/src/utils/shell-env.ts [hunk 6]
```diff
@@ -6,8 +6,18 @@
...
+  'https_proxy',
+  'HTTPS_PROXY',
+  'no_proxy',
+  'NO_PROXY',
+  'all_proxy',
+  'ALL_PROXY'
+] as const;
+
 const fetchShellEnv = async (): Promise<Record<string, string>> => {
-  // Windows handles environment variables differently - skip
```
Comment (helloanoop) [inline packages/bruno-requests/src/utils/shell-env.ts:10]: why remove this useful comment ? its the very least saying that the reason we are returning empty object is because `Windows handles environment variables differently` although I wish more context was commented

Comment (abhishekp-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts:10]: My bad I actually tried to explore handling proxy for windows and was working on it but then reverted the changes since it was taking too much time and also it was out of scope for this ticket. 
will add the comment back

File: packages/bruno-requests/src/utils/shell-env.spec.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
-import path from 'path';
 import { initializeShellEnv } from './shell-env';
+import path from 'path';
```
Comment (sanish-bruno) [inline packages/bruno-requests/src/utils/shell-env.spec.ts:1]: rm this, i think don't need to reorder, keep the such imports always at the top, like library imports, followed by imports from local files

File: packages/bruno-requests/src/utils/shell-env-proxy-refresh.spec.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+describe('refreshShellEnvProxyVars — shell inheritance', () => {
+  beforeEach(() => {
+    proxyVarsSeenByShell = {};
+    mockShellEnvResult = {};
+    for (const key of PROXY_ENV_KEYS) {
+      delete process.env[key];
+    }
+  });
+
+  afterEach(() => {
```
Comment (sanish-bruno) [inline packages/bruno-requests/src/utils/shell-env-proxy-refresh.spec.ts]: do we need to do the delete in both afterEach and beforeEach? , i think we can drop afteEach

File: packages/bruno-requests/src/network/system-proxy/get-system-proxy.spec.ts [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+ * To avoid running platform-restricted commands natively (which would crash tests run
+ * on mismatched host platforms or CI/CD pipelines), this file uses a dynamic sandbox runtime via
+ * `loadForPlatform(platform)`.
+ * * Instead of rigid invocation-sequence queues, platform CLI utilities are mocked using argument-aware
+ * routing logic (`execFile.mockImplementation`). This guarantees that structural changes to the internal
+ * execution sequence inside the proxy resolver will not cause brittle test breaks, keeping tests
+ * strictly tied to structural string contract parsing and logical compliance.
+ */
+
+const PROXY_ENV_KEYS = [
```
Comment (sanish-bruno) [inline packages/bruno-requests/src/network/system-proxy/get-system-proxy.spec.ts]: @abhishekp-bruno we have the option to split this up, may be even separate PR that handles cross-platform integration tests, better to add them to `network/system-proxy/index.spec.js`, can consider doing it.

and trim `refresh.spec.ts` to just the two cases that actually exercise `efreshShellEnvProxyVars` end-to-end (macOS env-wins-over-system, and env-cleared-after-refresh).

Also should we add a no-op test for `refreshShellEnvProxyVars` to assert the windows behaviour?

**Review by sanish-bruno — CHANGES_REQUESTED** [review]

File: packages/bruno-requests/src/utils/shell-env-proxy-refresh.spec.ts (hunk 8 above)
Comment (abhishekp-bruno) [reply packages/bruno-requests/src/utils/shell-env-proxy-refresh.spec.ts]: removed afterEach

File: packages/bruno-requests/src/utils/shell-env.spec.ts (hunk 7 above)
Comment (abhishekp-bruno) [reply packages/bruno-requests/src/utils/shell-env.spec.ts:1]: This happened while resolving conflicts have updated this

File: packages/bruno-requests/src/utils/shell-env.ts (hunk 3 above)
Comment (sanish-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts:86]: Add the early return cause we unconditioanlly deletes all the process.envs without considering whether they are from the shell or not, the refreshShellEnv can also be no-op in windows, just like fetchShellEnv. to avoid any possible regressions due to deletion @abhishekp-bruno

File: packages/bruno-requests/src/utils/shell-env.ts [hunk 10]
```diff
@@ -39,3 +50,36 @@ export const initializeShellEnv = async (): Promise<Record<string, string>> => {
...
+  // would strip process.env proxy vars (registry-propagated, launcher-set, or inherited
+  // from the parent shell) without any means to restore them. Preserve the additive-only
+  // invariant that initializeShellEnv maintains on Windows.
+  if (process.platform === 'win32') {
+    return {};
+  }
+
+  // Clear stale proxy vars first so shell-env does not inherit them into the
+  // login shell subprocess (removed .zshrc exports would otherwise persist).
+  for (const key of PROXY_ENV_KEYS) {
```
Comment (sanish-bruno) [inline packages/bruno-requests/src/utils/shell-env.ts:97]: Since we are actually deleting all the PROXY_ENVs here, then calling the `fetchShellEnv`, if any error the happens in `fetchShellEnv`, as it returns `{}` within the catch, refresh will remove all the existing values. Edge case, but something to be mindful about.  @abhishekp-bruno cc: @vijayh-bruno 

something like this can work, may also need to update the failing test accordingly

```js
  // New helper that surfaces failure. Keep fetchShellEnv as-is for initializeShellEnv's additive path.
  const tryFetchShellEnv = async (): Promise<Record<string, string> | null> => {
    if (process.platform === 'win32') return {};
    try {
      const { shellEnv } = await import('shell-env');
      return await shellEnv();
    } catch {
      return null;
    }
  };

  export const refreshShellEnvProxyVars = async (): Promise<Record<string, string>> => {
    if (process.platform === 'win32') return {};

    const snapshot: Record<string, string | undefined> = {};
    for (const key of PROXY_ENV_KEYS) {
      snapshot[key] = process.env[key];
      delete process.env[key];
    }

    const shellEnvVars = await tryFetchShellEnv();

    if (shellEnvVars === null) {
      // Subprocess failed — restore prior values rather than leave the user unproxied.
      for (const key of PROXY_ENV_KEYS) {
        if (snapshot[key] !== undefined) process.env[key] = snapshot[key];
      }
      return {};
    }

    for (const key of PROXY_ENV_KEYS) {
      const value = shellEnvVars[key];
      if (value) process.env[key] = value;
    }
    return shellEnvVars;
  };

```

**Review by sanish-bruno — CHANGES_REQUESTED** [review]

**Review by sanish-bruno — APPROVED** [review]

**Review by vijayh-bruno — DISMISSED** [review]

File: packages/bruno-requests/src/utils/shell-env.ts (hunk 10 above)
Comment (sid-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts:97]: @abhishekp-bruno this is still valid in a situation where the race function causes this one to just be left in the timeout, we'll end up with a partial snapshot which isn't something we want. 

Needs to be addressed.

**Review by sid-bruno — CHANGES_REQUESTED** [review]

File: packages/bruno-requests/src/utils/shell-env.ts (hunk 3 above)
Comment (abhishekp-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts:86]: We are skipping windows here because everything related to windows proxy settings is handled by the system proxy resolver i.e getSystemProxy()

File: packages/bruno-requests/src/utils/shell-env.ts (hunk 10 above)
Comment (abhishekp-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts:97]: I have updated the code to handle the race condition inside the function itself

Comment (sid-bruno) [reply packages/bruno-requests/src/utils/shell-env.ts:97]: ~~Timeout shouldn't be in bruno-requests, it shouldn't be holding business and rate limiting in here but should be in the app / cli layer only. you can just move the delete logic to be run after fetching has finished instead.~~

The rest looks good, let me discuss with @bijin-bruno and get back

**Comment by bijin-bruno** [comment]
Closed as merged via https://github.com/usebruno/bruno/pull/8535
